test/spec/support/compile-config-spec.js

```diff
@@ -69,79 +69,97 @@ describe('support/compile-config', function() {
 
   describe('plugins', function() {
 
-    it('should import exported and non-exported', async function() {
-
-      // given
-      const resolver = new NodeResolver({
-        require: function(path) {
-          if (path === 'bpmnlint-plugin-foreign') {
-            return {
-              configs: {
-                recommended: {
-                  rules: {
-                    'exported': 'error',
-                    'non-exported': 'error'
+    describe('foreign', function() {
+
+      it('should import exported and non-exported', async function() {
+
+        // given
+        const resolver = new NodeResolver({
+          require: function(path) {
+            if (path === 'bpmnlint-plugin-foreign') {
+              return {
+                configs: {
+                  recommended: {
+                    rules: {
+                      'exported': 'error',
+                      'non-exported': 'error'
+                    }
                   }
+                },
+                rules: {
+                  'exported': () => {}
                 }
-              },
-              rules: {
-                'exported': () => {}
-              }
-            };
+              };
+            }
+
+            throw new Error('not found');
           }
+        });
 
-          throw new Error('not found');
-        }
-      });
 
+        // when
+        const code = await compileConfig({
+          extends: 'plugin:foreign/recommended'
+        }, resolver);
 
-      // when
-      const code = await compileConfig({
-        extends: 'plugin:foreign/recommended'
-      }, resolver);
+        // then
+        expect(code).to.contain('import exports_0 from \'bpmnlint-plugin-foreign\'');
+        expect(code).to.contain('cache[\'bpmnlint-plugin-foreign/exported\'] = exports_0.rules[\'exported\']');
```

Comment:
This cannot be tree-shaken, so what will end up in the final bundle will always be _all_ exported rules.

What we could do:

* Stick to our current _virtual_ style of declaring things: Instead of requiring a particular rule, you define the relative path to it:

#### Not this

```
module.exports = {
  rules: {
    baz: require('./foo/bar/baz')
  }
};
```

#### But this

```
// contract: rule path is relative to plugin root folder
module.exports = {
  rules: {
    baz: './foo/bar/baz'
  }
};
```

#### Why?

This allows us to generate more efficient bundles:

```
import rule_0 from 'bpmnlint-plugin-foreign/foo/bar/baz';

cache['bpmnlint-plugin-foreign/exported'] = rule_0;
```

Reply:
Are you proposing changing that in both `node-resolver` and `compile-config`? Should we support both strings and functions or strings only? ESLint seems to only support functions so they don't seem to care about tree shaking.

Reply:
Yes. Eslint is Node only, it is a pain to get running in C8.

Reply:
Alternatively:

* let's test the current solution with bundlers and see if these are smart enough
* decide we don't want tree shaking as a feature

Reply:
I can have a look and wrap up this PR next week.